# Conditionals and Iterables

&nbsp;


# 1- Python's Conceptual Hierarchy:

<span style='font-size:1.1em'>
    1- Programs are comprised of modules   
    2- Modules are comprised of statements  
    3- Statements are comprised of expressions   
    4- Expressions create and process objects</span>
&nbsp;

* An Expression is defined as a collection of symbols that jointly express a quantity. 2$\pi$r is an expression of the circumference of a circle. 
* Expressions produce a value however because statements can also be used to produce a value this distinction can become blurry at times.  

* expressions

In [1]:
'greetings'

'greetings'

In [2]:
[4,1,5,6,7]

[4, 1, 5, 6, 7]

In [3]:
# an example of a statement to create an expression.
# this expression is formally known as a list comprehension

[x**2 for x in [4,1,5,6]]

[16, 1, 25, 36]

* statements

In [8]:
str_1 = 'greetings'

r = [x**2 for x in [4,1,5,6]]

[16, 1, 25, 36]

* Everything that we have encountered up to this point has mostly been Python expressions or statements used to create other statments.
* A statement can be thought of as the smallest standalone element of an imperative programming language. 
* Famous Python statements include `print`, `if/else/elif`, `for/else`, `while/else`, `pass/break/continue`, `def`, `return`, `from`, `import`, `class` and `del`

&nbsp;

&nbsp;


# 2- Conditional Statements in Python
_________________________________
* conditional statements are about decision making   
* Syntax:     
<br>  
`if condition1:    
    statement 1    
elif condition2:  
    statement 2    
else:  
    statement 3`

In [16]:
list_1 = [2,3,5,6]
list_1 = [3,6,5,4,1]
list_1 = [5,1,9]

In [17]:
if len(list_1) == 4:
    print(list_1*2)
elif len(list_1) > 4:
    print('the list has number of more than 4 elements')
else:
    print([x**2 for x in list_1])

[25, 1, 81]


* a simple `if/else` statement can be written efficiently as a single line expression

In [18]:
print('list_1 has 4 elements') if len(list_1)==4 else print('no info on length of list_1')

no info on length of list_1


In [20]:
y = 12
x = 7 if y >=4 else 9
print('x = {}'.format(x))

x = 7


In [ ]:
y=3

In [21]:
print('x = {}'.format(7 if y >=4 else 9))

x = 7


* an `if x:` not followed with any condition is equivalent to `if x == True:` or `if x == ''`   
* `if` statements, `for` loops and `def` methods are very sensitive to indentations.   
* indentation defines code block boundaries in Python methods and statements.  

In [22]:
x = 1
if x:         # this checks if x is empty
    y = 2     # this checks if y is empty
    if y:
        print('block2')  # this is the result of condition 2
    print('block1')      # this is result of condition 1
print('block0')          # this is outside the if statement

block2
block1
block0


In [23]:
x = ''
if x:
    y = True
    if y:
        print('block2')
    print('block1')
print('block0')

block0


* an `if` statement can have multiple `elif` statements.  

In [24]:
BMI = 26

if BMI < 18.5: 
    print('bmi of {} is considered underweight'.format(BMI))
elif BMI >= 18.5 and BMI < 24.9:
    print('bmi of {} is considered healthy'.format(BMI))
elif BMI >=24.9 and BMI < 29.9:
    print('bmi of {} is considered overweight'.format(BMI))
elif BMI >=29.9 and BMI < 39.9:
    print('bmi of {} is considered obese'.format(BMI))
else:
    print('bmi of {} is considered extemely obese'.format(BMI))
    

bmi of 26 is considered overweight


* `if not` statement is used most conveniently when the condition is checking an object attribute.
* `if not` is very useful in expressions and list comprehensions. 
       
       
* below:    
`if not word.count(sub, 0, len(word))` is more succint than `if word.count(sub, 0, len(word)) == 0`

In [31]:
word = 'epanalepsis'
sub = 'eps'
print('{} does not occur'.format(sub))\
if not word.count(sub, 0, len(word)) else print('{} occurs'.format(sub)) 

eps occurs


&nbsp;

&nbsp;

# 3- Loops and Iterators in Python
_______________________
* two main looping constructs `for` and `while` statements.   
* a number of supporting statements that can be nested inside loops:      
    1- `break` jumps out of the closest enclosing loop (requires a nested if statement).     
    2- `continue` jump to the top of the closest enclosing loop (requires a nested if statement).   
    3- `else` block, runs if the loop is exited normally.      
    4- `pass` an empty statement placeholder (most useful when building a looping construct).       
       

## 3.1 `while` statement

* `while` statements often involves a value that changes after each iteration.

`while test:
    statements`
  
 *with an optional else statement*    
  
`while test:
    statements
else:
    statements`

In [32]:
echo = 'testing'
while len(echo)>2:
    print(echo, end=' ')
    echo = echo[1:]

testing esting sting ting ing 

* a `while-else` statement is a statement that executes `else` when the `while` loop is exits normally. 

In [33]:
echo = 'testing'
while len(echo)>2:
    print(echo, end=' ')
    echo = echo[1:]
else:
    print('\nsmaller subsets not part of echo')

testing esting sting ting ing 
smaller subsets not part of echo


* `continue` statement will skip the expression of the `while` loop upon satisfying a condition.

In [34]:
x = 12
while x > 1:
    x-=1                      # x-=1 == x = x-1
    if x % 2 != 0: continue   # this will skip value of x 11,9,7, and 5
    print((x,x/2), end=' ')

(10, 5.0) (8, 4.0) (6, 3.0) (4, 2.0) (2, 1.0) 

in the example above since the statement will decrement `x` before checking if it is even the first value of x is not assessed.  

* the `break` statement breaks the loop once a condition is satisfied

In [35]:
employee_name=[]

In [36]:
while True:
    name = input('Enter Name: ')
    if name == 'stop': break
    employee_name.append(name)

Enter Name: ARROW
Enter Name: stop


In [37]:
employee_name

['ARROW']

&nbsp;

* fibonacci sequence can be created using a `while` statement. 
* every number after the first two is the sum of the two preceeding oncess: $$ F_n = F_{n-1} + F_{n-2} $$

<span style='color:blue'>write a simple while loop that creates a list of the first 21 numbers of the fibonnacci sequence. start by defining a core list of `[0,1]` or the first two numbers `x=0` and `y=1`</span>

In [23]:
#skipped code
fiblist=[0,1]
x=0
y=1
while len(fiblist)<21 :
    x,y=y,y+x
    fiblist.append(y)
    
print(fiblist)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765]


21

In [ ]:
#My codes were written in a list, but the instructor's code 

here's the first 21 numbers in the sequence

`0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 6765`

&nbsp;


## 3.2 `for` statement   

`for target in object:
    statements`
  
 *with an optional else statement*    
  
`for target in object:
    statements
else:
    statements`

* strings in Python are iterable objects types

In [9]:
s = 'this is a string'

for c in s:
    print(c+'-', end='')

t-h-i-s- -i-s- -a- -s-t-r-i-n-g-

In [10]:
for c in s:
    print(ord(c), end=' ')

116 104 105 115 32 105 115 32 97 32 115 116 114 105 110 103 

this can be written more succinctly as:    

In [11]:
[ord(c) for c in s]

[116, 104, 105, 115, 32, 105, 115, 32, 97, 32, 115, 116, 114, 105, 110, 103]

(above is an example of an expression called a list comprehension)
&nbsp;

&nbsp;

* `for` loops can also act in the capacity of a `while` statement to repeatadly evaluate an expression

In [12]:
import numpy as np

In [13]:
# nothice that i is not being utilized in the expression
for i in range(100):
    print(chr(int((round(np.random.rand(),2)*100))),end=' ')

 I  I D !  L M M J  +  , ? L  	 H _  _ V  ]  " % X " \ I T & W  J > 0    ) 1 L = 

* `for` statements can be nested

In [24]:
a, b = range(2,8), range(1,3)
a_pow_b = [] #empty list to append new elements to. 

for i in a:
    for j in b:
        a_pow_b.append(pow(i,j))
        
print(a_pow_b)

[2, 4, 3, 9, 4, 16, 5, 25, 6, 36, 7, 49]


* keys and values of a dictionary are iterable, use a for loop to iterate thru a dictionary items:

In [ ]:
dict_4 = {1985:'Cycle of Warewolf',
          1996:{'The Green Miles':{2:'The Mouse on the Mile',1:'The Two Dead Girls'}, 3:'Coffery\' Hands',
               5:'Night Journey',6:'Coffery on the Mile',4:'The Bad Death of Edward Delacriox'},
          1991:{1:'The Stand',2:'The Dark Tower:The Waste Lands'},
          1987:{1:'The Eyes of the Dragon',2:'The Dark Tower:The Drawing of the tree',3:'Misery',4:'The Tommyknockers'}}

In [ ]:
# in this example we call key the attribute of dict_4
for key in dict_4:
    print(key,' => ',dict_4[key], end='\n\n')

In [ ]:
# in this example x and y represent the key and value assigned to the dict_4.items()
# recall that dict_4.items() returns a list of (key,value) tuples that we capture using multi-assignment

for (x,y) in dict_4.items():
    print(x,' ==> ',y, end='\n\n')

### 3.2.1 the `enumerate()` function:

* syntax: `enumerate(iterable, start=0)`    
* by itself the `enumerate()` functions returns a tuple containing an (index, value) pair starting at defined `start`. 

In [7]:
# a sample from a normally distribute random variable.
norm_1 = [ 
        0.85567338,  0.26152381, -0.73251952, -0.19386771, -0.02788016,
       -0.37545653,  0.86448105,  0.14906788,  0.20980621,  0.01462341,
       -1.16974472,  0.3044212 ,  0.10179421,  0.5986373 , -0.153957  ,
       -0.53779243, -0.64934128, -0.18870407,  0.36207197, -0.87953644,
        1.5002981 ,  0.09291734, -0.22792116, -0.00722612, -0.01658935,
        0.67512695,  0.21235187, -0.98757662,  0.03776504,  1.72908671,
       -0.47101163, -0.33828453, -0.56690145,  0.22863732, -0.11394473,
        0.57888681,  0.41353363,  0.41353225, -0.11699942, -0.28788474,
        0.47052836,  1.31212803,  0.00881542,  0.33294558, -0.17610327,
        0.07685088,  0.13075439, -0.44796947, -0.82784466,  0.48043545,
        0.57840025,  0.76581154,  0.15029234,  0.18670172, -0.45963916,
       -0.29813299,  0.863169  ,  1.56515904,  1.42559799,  0.24279349,
       -0.46985923,  0.11271094, -0.10062158,  0.53074919,  1.17386216,
        0.3249393 ,  0.10828674, -0.0398253 , -0.532675  ,  0.2248118 ,
        0.34462597,  0.68117081, -0.66904952, -0.64081592, -0.79508548,
        1.15430465,  0.15814982,  0.56552673, -0.51990009,  0.27117282,
       -0.09115771,  0.19376023, -0.51766469,  0.24680429, -0.8824784 ,
        0.21518327,  0.14214064, -0.08711938, -0.52234801, -0.1186766 ,
        0.27748961, -0.12939926,  0.20221867, -0.11496247, -0.25483535,
        0.33208192,  0.53052014,  0.48706197,  1.01395269, -0.32689196]

In [ ]:
# we must list the enumerated object to 
list(enumerate(norm_1))[:10]

* when `enumerate()` is used within a `for` statement both the object and its index can be captured.   
* this is useful when capturing the index of an object becomes necessary.   

In [2]:
word_list = ['foo', 'bar', '423','gronk', 'hello kitty', 'sling', 'drag', '8' ]

In [3]:
# produces a tuple of (index, value)
digits = []

for loci, word in enumerate(word_list):
    if word.isdigit():digits.append((loci, word))
        
digits

[(2, '423'), (7, '8')]

&nbsp; 

<span style='color:blue'> write a for-statement that returns the index and values from <U>norm_1</U> that are less than 2 standard deviations from the mean </span> 

In [4]:
std = 0.57233146076364871
mean = 0.10419949169999999

In [13]:
#skipped code
vec = [] #use vec to append new value pairs to

for index, val in enumerate(norm_1):
    if mean-2*std>val or mean+2*std<val:
        vec.append((index,val))
    
vec      

[(10, -1.16974472),
 (20, 1.5002981),
 (29, 1.72908671),
 (41, 1.31212803),
 (57, 1.56515904),
 (58, 1.42559799)]

print looks like this 


`[(10, -1.16974472), (20, 1.5002981), (29, 1.72908671), (41, 1.31212803), (57, 1.56515904), (58, 1.42559799)]`

&nbsp;

### 3.2.2 `for-else` statement and `for-continue` statement  
* not very common but useful when required. 
* `for-else` statement is executed as the loop exits normally

In [14]:
s = 'this has to be a string'

for ch in s:
    print(ch+'-',end='')  # successful execution of the for statement carried out the else statemet
else:
    print('\n\nran out of letters!')

t-h-i-s- -h-a-s- -t-o- -b-e- -a- -s-t-r-i-n-g-

ran out of letters!


* `for-continue` statement continues with the next iteration of the loop (skipping the current iteration) upon a condition. 

In [15]:
my_list = [4,10,1,-1,7,12,-3,-6,2]

for index, num in enumerate(my_list):
    if num < 0: continue
    my_list[index] = num**2
else:
    print('done')

done


In [ ]:
my_list

&nbsp; 

### 3.2.3 list comprehensions   

* list comprehensions are expressions that are written in square brackets because they construct a new list.  
* we have sporadically come across list comprehensions before. 
* general syntax  
* the concept of comprehension in python makes iterable statements very powerful and succinct

`[do_something(element)` **`for`** `element` **`in`** `object]`


`[do_something(element)` **`for`** `element` **`in`** `object` **`if`** `condition` `]`

&nbsp;

<span style="color:blue">use a list comprhension to create a new list that is modulous `%` of `my_list` by 3</span>

In [16]:
my_list = [5,48,6,71,6,18,3,53,0,9]

In [22]:
#it will return a new list with all the elements in my_list
[i%3 for i in my_list]

[2, 0, 0, 2, 0, 0, 0, 2, 0, 0]

In [23]:
for i in my_list:
    print(i%3)

2
0
0
2
0
0
0
2
0
0


&nbsp;

read the file `keats_tabs.txt` into python

In [24]:
poem = tuple(open('data/keats_tabs.txt','r'))
poem

('\t\t\tDeep in the shady sadness of a vale\t\t\t \n',
 '\t\t\tFar sunken from the healthy breath of morn,\t\t\t \n',
 "\t\t\tFar from the fiery noon, and eve's one star,\t\t\t \n",
 "\t\t\tSat gray-hair'd Saturn, quiet as a stone,\t\t\t \n",
 '\t\t\tStill as the silence round about his lair;\t\t\t \n',
 '\t\t\tForest on forest hung about his head \t\t\t\n',
 '\t\t\tLike cloud on cloud. No stir of air was there,\t\t\t \n',
 "\t\t\tNot so much life as on a summer's day\t\t\t \n",
 "\t\t\tRobs not one light seed from the feather'd grass,\t\t\t \n",
 '\t\t\tBut where the dead leaf fell, there did it rest.\t\t\t \n',
 '\t\t\tA stream went voiceless by, still deadened more\t\t\t \n',
 '\t\t\tBy reason of his fallen divinity \t\t\t\n',
 "\t\t\tSpreading a shade: the Naiad 'mid her reeds\t\t\t \n",
 "\t\t\tPress'd her cold finger closer to her lips.\t\t\t")

In [25]:
print(poem[0])

			Deep in the shady sadness of a vale			 



&nbsp;

<span style="color:blue"> using the string instance method `.strip()` remove the tabs from each line</span>

In [28]:
#skipped code
poem=[line.strip() for line in poem]
poem

['Deep in the shady sadness of a vale',
 'Far sunken from the healthy breath of morn,',
 "Far from the fiery noon, and eve's one star,",
 "Sat gray-hair'd Saturn, quiet as a stone,",
 'Still as the silence round about his lair;',
 'Forest on forest hung about his head',
 'Like cloud on cloud. No stir of air was there,',
 "Not so much life as on a summer's day",
 "Robs not one light seed from the feather'd grass,",
 'But where the dead leaf fell, there did it rest.',
 'A stream went voiceless by, still deadened more',
 'By reason of his fallen divinity',
 "Spreading a shade: the Naiad 'mid her reeds",
 "Press'd her cold finger closer to her lips."]

In [29]:
print(poem[0], poem[1], sep='\n')

Deep in the shady sadness of a vale
Far sunken from the healthy breath of morn,


In [30]:
[line.upper() for line in poem]

['DEEP IN THE SHADY SADNESS OF A VALE',
 'FAR SUNKEN FROM THE HEALTHY BREATH OF MORN,',
 "FAR FROM THE FIERY NOON, AND EVE'S ONE STAR,",
 "SAT GRAY-HAIR'D SATURN, QUIET AS A STONE,",
 'STILL AS THE SILENCE ROUND ABOUT HIS LAIR;',
 'FOREST ON FOREST HUNG ABOUT HIS HEAD',
 'LIKE CLOUD ON CLOUD. NO STIR OF AIR WAS THERE,',
 "NOT SO MUCH LIFE AS ON A SUMMER'S DAY",
 "ROBS NOT ONE LIGHT SEED FROM THE FEATHER'D GRASS,",
 'BUT WHERE THE DEAD LEAF FELL, THERE DID IT REST.',
 'A STREAM WENT VOICELESS BY, STILL DEADENED MORE',
 'BY REASON OF HIS FALLEN DIVINITY',
 "SPREADING A SHADE: THE NAIAD 'MID HER REEDS",
 "PRESS'D HER COLD FINGER CLOSER TO HER LIPS."]

&nbsp;

* a filtered list comprehension contains a conditional. 
* this list comprehension picks lines longer than 45 characters and reorders the characters alphabetically

In [34]:
#return the first letter of each line
''.join(sorted(list(poem[0])))

'       Daaaaddeeeeefhhilnnopsssstvy'

In [33]:

[''.join(sorted(list(line))) for line in poem if len(line)>45]

['         ,.LNaaccddeeefhiiikllnooooorrrssttuuw',
 "        ',Raabddeeeeeeffgghhhilmnnoooorrrsssstttt",
 '         ,.Baaddddeeeeeeeeeffhhhiilllrrrstttttuw',
 '       ,Aaabcdddeeeeeeeeiilllmmnnoorrsssstttvwy']

* list comprehensions can accommodate multiple `for` statements

In [35]:
x , y , z = range(2,6), range(3,10), range(4,5)

[(a**b)/c for a in x for b in y for c in z]

[2.0,
 4.0,
 8.0,
 16.0,
 32.0,
 64.0,
 128.0,
 6.75,
 20.25,
 60.75,
 182.25,
 546.75,
 1640.25,
 4920.75,
 16.0,
 64.0,
 256.0,
 1024.0,
 4096.0,
 16384.0,
 65536.0,
 31.25,
 156.25,
 781.25,
 3906.25,
 19531.25,
 97656.25,
 488281.25]

In [36]:
print([a + b for a in x for b in y])

[5, 6, 7, 8, 9, 10, 11, 6, 7, 8, 9, 10, 11, 12, 7, 8, 9, 10, 11, 12, 13, 8, 9, 10, 11, 12, 13, 14]


In [37]:
print([a*b for a in x for b in ['s','p','a','m']])

['ss', 'pp', 'aa', 'mm', 'sss', 'ppp', 'aaa', 'mmm', 'ssss', 'pppp', 'aaaa', 'mmmm', 'sssss', 'ppppp', 'aaaaa', 'mmmmm']


<span style='color:blue'>write a list comprehension that will print an (index, value) pair for values that are less than 2x standard deviation from the mean.</span>


In [ ]:
#skipped code


In [38]:
#skipped code
[(index, val) for index,val in enumerate(norm_1) if (val > mean+2*std) or (val < mean-2*std)]

[(10, -1.16974472),
 (20, 1.5002981),
 (29, 1.72908671),
 (41, 1.31212803),
 (57, 1.56515904),
 (58, 1.42559799)]

print result looks like this    


`[(20, 1.5002981),    
 (29, 1.72908671),    
 (41, 1.31212803),   
 (57, 1.56515904),   
 (58, 1.42559799)]`   

&nbsp;

finally, when we write a **condidional list comprehension** with multiple condidions the syntax becomes different   
      
with a single condition the general syntax is:  

`[do_something(element)`**`for`**`element`**`in`**`object`**`if`**`condition]`


with multiple **`if`** conditions the the syntax handles all conditions prior to the **`for`** expression: 

`[a `**`if`**`i==cond_1 `**`else`**`b`**`if`**`i==cond_2`**`else`**`c`**`if`**`i==cond_3`**`else`**`d`**`for`**`i in list]`

In [ ]:
weight_list = [11,26,18,19,31,12,41]

In [42]:
["underweight" if i < 18.5 else "healthy" 
 if i>=18.5 and i<24.5 else "overweight" 
 if i>=24.5 and i<29.9 else "obese" 
 if i>=29.9 and i<39.9 else "extramely obses" for i in weight_list] 

['underweight',
 'overweight',
 'underweight',
 'healthy',
 'obese',
 'underweight',
 'extramely obses']

&nbsp; 

### 3.2.4 dictionary comprehensions   

* similar to a list comprehensions a dictionary comprehension is created using comprehension syntax enclosed with curly brackets.    

In [43]:
my_dict = {x: x**2 for x in range(11) if x % 2 == 0 }
print(my_dict)

{0: 0, 2: 4, 4: 16, 6: 36, 8: 64, 10: 100}


* nested `for` statements in dictionary comprehension do not work as we expected (as they do in list comprehensions).

In [44]:
[(y,y*x) for y in 'LOL' for x in list(range(11)) ]

[('L', ''),
 ('L', 'L'),
 ('L', 'LL'),
 ('L', 'LLL'),
 ('L', 'LLLL'),
 ('L', 'LLLLL'),
 ('L', 'LLLLLL'),
 ('L', 'LLLLLLL'),
 ('L', 'LLLLLLLL'),
 ('L', 'LLLLLLLLL'),
 ('L', 'LLLLLLLLLL'),
 ('O', ''),
 ('O', 'O'),
 ('O', 'OO'),
 ('O', 'OOO'),
 ('O', 'OOOO'),
 ('O', 'OOOOO'),
 ('O', 'OOOOOO'),
 ('O', 'OOOOOOO'),
 ('O', 'OOOOOOOO'),
 ('O', 'OOOOOOOOO'),
 ('O', 'OOOOOOOOOO'),
 ('L', ''),
 ('L', 'L'),
 ('L', 'LL'),
 ('L', 'LLL'),
 ('L', 'LLLL'),
 ('L', 'LLLLL'),
 ('L', 'LLLLLL'),
 ('L', 'LLLLLLL'),
 ('L', 'LLLLLLLL'),
 ('L', 'LLLLLLLLL'),
 ('L', 'LLLLLLLLLL')]

In [45]:
my_dict = {y: y*x for y in 'ABCDEFGHIGJ' for x in list(range(11)) }
print(my_dict)

{'A': 'AAAAAAAAAA', 'B': 'BBBBBBBBBB', 'C': 'CCCCCCCCCC', 'D': 'DDDDDDDDDD', 'E': 'EEEEEEEEEE', 'F': 'FFFFFFFFFF', 'G': 'GGGGGGGGGG', 'H': 'HHHHHHHHHH', 'I': 'IIIIIIIIII', 'J': 'JJJJJJJJJJ'}


In [46]:
{x + y: (ord(x), ord(y)) for x in 'lov' for y in 'abc'}

{'la': (108, 97),
 'lb': (108, 98),
 'lc': (108, 99),
 'oa': (111, 97),
 'ob': (111, 98),
 'oc': (111, 99),
 'va': (118, 97),
 'vb': (118, 98),
 'vc': (118, 99)}

&nbsp; 
### Additional iterables

* `zip()` is a new less famous iterator introduced in python 3.X include 
* `zip()` can combine elements from multiple tuples of the same length according to their order and produce a list of tuples.
* `zip()` is a size optimized construct, even as the length of the sequences grows the size of a zip object remains constant. 

In [47]:
zip_1 = zip(('Jon','Jamie','Daenerys'),(55000, 65000, 72000), ('312-885-1656', '312-995-1112', '857-400-6400'))

zip_2 = zip(('Jon','Jamie','Daenerys','Jorah'),('Snow','Lannister','Targaryen','Mormont'),(55000, 65000, 72000, 61000), 
            ('312-885-1656', '312-995-1112', '857-400-6400', '800-265-3338'))

In [48]:
a = [('Jon', 55000, '312-885-1656'),
 ('Jamie', 65000, '312-995-1112'),
 ('Daenerys', 72000, '857-400-6400')]

In [49]:
b = [('Jon', 'Snow', 55000, '312-885-1656'),
 ('Jamie', 'Lannister', 65000, '312-995-1112'),
 ('Daenerys', 'Targaryen', 72000, '857-400-6400'),
 ('Jorah', 'Mormont', 61000, '800-265-3338')]

In [50]:
a, b

([('Jon', 55000, '312-885-1656'),
  ('Jamie', 65000, '312-995-1112'),
  ('Daenerys', 72000, '857-400-6400')],
 [('Jon', 'Snow', 55000, '312-885-1656'),
  ('Jamie', 'Lannister', 65000, '312-995-1112'),
  ('Daenerys', 'Targaryen', 72000, '857-400-6400'),
  ('Jorah', 'Mormont', 61000, '800-265-3338')])

In [51]:
#zip objects will never exceed the size of 64 bytes
from sys import getsizeof
getsizeof(zip_1), getsizeof(zip_2), getsizeof(a), getsizeof(b)

(64, 64, 88, 96)

In [58]:
#list shows the contents once, and after that, it'll be empty
list(zip_1)

[]

In [59]:
#iterators map, zip, generator

&nbsp;

&nbsp;

### Additional numeric operators in Python     

* the following operators are very useful in the context of iterable statements

+= add and assign   
-= subtract assign   
\*= multiply and assign    
/= divide and assign     
\**= exponentiate and assign  
 

In [60]:
l = 2

for i in range(1,7):
    l**=i
    print(l)

2
4
64
16777216
1329227995784915872903807060280344576
5515652263101987298728728207430913795608113109085112352897269396216198887424215820128660001943808587833784893551335930816647064191168732319583111500951066614122648616177179922993422016587311577585463592732098692120576


In [ ]:
l = 3985

for i in range(1,7):
    l//=i
    print(l)